File: master/master/page/bom/bom.py
import frappe 
from frappe import _
import frappe

@frappe.whitelist(allow_guest=True)
def get_bom_list():
    bom_list = frappe.get_all('BOM List', fields=['bom_id', 'description', 'mc_types'])

    ht = "<table>"
    ht += "<tr>"

    # Generate headings for BOM list fields
    for i in bom_list[0].keys():
        ht += f'<th class="bg-success">{i}</th>'
    ht += "<th class='bg-success'>Child Table</th>"
    ht += "</tr>"
    
    # Iterate through BOM list items
    for bom_item in bom_list:
        ht += "<tr>"
        
        # Populate BOM list data
        for key, value in bom_item.items():
            ht += f'<td>{value}</td>'
        
        # Call get_child_table() with the bom_id for each BOM item
        child_table_data = get_child_table(bom_item['bom_id'])
        
        # Populate child table data
        ht += f'<td>{child_table_data}</td>'
        
        ht += "</tr>"
    
    ht += "</table>"
    return ht

@frappe.whitelist(allow_guest=True)
def get_child_table(bom_id):
    child_table = frappe.get_all('BOM child table', filters={'parent': bom_id}, fields=['item_code', 'description', 'qty_used', 'uom'])
    
    if child_table and len(child_table) > 0:
        ht = "<table>"
        ht += "<tr>"
        
        # Generate headings for child table fields
        for key in child_table[0].keys():     
            ht += f'<th class="bg-success">{key}</th>'
        ht += "</tr>"
        
        # Populate child table data
        for child_item in child_table:
            ht += "<tr>"
            for _, value in child_item.items():
                ht += f'<td>{value}</td>'
            ht += "</tr>"
        
        ht += "</table>"
        return ht
    else:
        return ""  # Return an empty string if child_table is empty



import json
@frappe.whitelist()
def save_bom_list(data):
    data = json.loads(data)
    print(data)
    bom_id = data['bom_id']
    description = data['description']
    mc_types = data['mc_types']
    bom_items=data['bom_items']


    bom_list = frappe.new_doc("BOM List")
    bom_table = frappe.new_doc("BOM child table")
    bom_list.bom_id = bom_id
    bom_list.description = description
    bom_list.mc_types = mc_types
    bom_list.insert(ignore_permissions=True)

    for bom_item in bom_items:
        bom_table.item_code = bom_item.get("item_code")
        bom_table.description =bom_item.get('child_description')
        bom_table.qty_used = bom_item.get("qty_used")
        bom_table.uom =bom_item.get("uom")
        bom_table.parent =bom_item.get("parent")
        bom_table.parenttype = bom_item.get("parenttype") 
        bom_table.insert(ignore_permissions=True)
    
    return "Bom list sucessfully submitted"

    

File: master/master/page/bom/bom.js
frappe.pages['bom'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'BOM List',
		single_column: true
	});
	$(frappe.render_template("bom", {})).appendTo(page.body);

	frappe.call({
		method:'master.master.page.bom.bom.get_bom_list',
		args:{},
		callback:function(response){
			var m = response.message;
			document.getElementById("bom").innerHTML=m;
			console.log(m);
		}
	});
	// frappe.call({
	// 	methoget_child_tabled:'master.master.page.bom.bom.get_child_table',
	// 	args:{
	// 		'bom_id': document.querySelector("#bom_id").value
	// 	},
	// 	callback:function(response){
	// 		var mes = response.message;
	// 		document.getElementById("child_bom").innerHTML=mes;
	// 		console.log(mes);
	// 	}
	// })


		var bomform = document.querySelector('#bomform');
		bomform.addEventListener("submit", function(event) {
			event.preventDefault();
	
			var formData = {
				'bom_id': document.querySelector("#bom_id").value,
				'description': document.querySelector('#description').value,
				'mc_types': document.querySelector('#mc_types').value,
				'bom_items':[{
					'item_code':document.querySelector('#item_code').value,
					'child_description':document.querySelector('#description1').value,
					'qty_used':document.querySelector('#qty_used').value,
					"uom":document.querySelector('#uom').value,	
					'parent':document.querySelector("#bom_id").value,
					'parenttype':"BOM List"
				}
				]

			};
			console.log("FormData", formData);
	
			frappe.call({
				method: 'master.master.page.bom.bom.save_bom_list',
				args: {
					'data': formData
				},
				callback: function(response) {
					console.log(response);
					if (response.message) {
						alert("You successfully submitted");
					} else {
						alert("Failed to create the BOM");
					}
				}
			});
		});	
}

//validate for avoid text and allow only numbers
function validateInput(event){
	var input = event.target;
	input.value = input.value.replace(/\D/g, '')
	if (event.which < 48 || event.which > 57) {
	event.preventDefault();
	
  }
  page.reset();
}

//add the row to the child tableFormData
function addRow() {
    var table = document.getElementById("myTable");
    var newRow = table.insertRow(-1); 
    newRow.innerHTML = `
        <td><input type="checkbox"></td>
        <td><input type="text" class="value" onkeypress="validateInput(event)"></td>
        <td><textarea class="value"></textarea></td>
        <td><input type="text" class="value" onkeypress="validateInput(event)"></td>
		<td><select class="value"><option>Number</option><option>Bar</option></select></td>
    `;
}


//delete the particular row in table
function deleteRow() {
	var table = document.getElementById("myTable").getElementsByTagName('tbody')[0];
	var rows = table.rows;
	for (var i = 0; i < rows.length; i++) {
		var row = rows[i];
		var chkbox = row.cells[0].getElementsByTagName('input')[0];
		if (chkbox.checked) {
			table.deleteRow(i);
			i--; // Adjust the index as rows are deleted
		}
	}
}

